crates/solvers/src/domain/dex.rs

```diff
@@ -0,0 +1,143 @@
+//! Various solver implementations rely on quoting APIs from DEXs and DEX
+//! aggregators. This domain module models the various types around quoting
+//! single orders with DEXs and turning swaps into single order solutions.
+
+use {
+    crate::{
+        domain::{eth, order, solution},
+        util::fmt,
+    },
+    ethereum_types::U256,
+    std::fmt::{Debug, Formatter},
+};
+
+/// An order for quoting with an external DEX or DEX aggregator. This is a
+/// simplified representation of a CoW Protocol order.
+#[derive(Debug)]
+pub struct Order {
+    pub sell: eth::TokenAddress,
+    pub buy: eth::TokenAddress,
+    pub side: order::Side,
+    pub amount: Amount,
+}
+
+impl Order {
+    /// Returns the order swapped amount as an asset. The token associated with
+    /// the asset is dependent on the side of the DEX order.
+    pub fn amount(&self) -> eth::Asset {
+        eth::Asset {
+            token: match self.side {
+                order::Side::Buy => self.buy,
+                order::Side::Sell => self.sell,
+            },
+            amount: self.amount.0,
+        }
+    }
+}
+
+impl From<order::Order> for Order {
+    fn from(value: order::Order) -> Self {
+        Self {
+            sell: value.sell.token,
+            buy: value.buy.token,
+            side: value.side,
+            amount: Amount(match value.side {
+                order::Side::Buy => value.buy.amount,
+                order::Side::Sell => value.buy.amount,
+            }),
+        }
+    }
+}
+
+/// A slippage tolerance (as a factor) that can be applied to token amounts.
+#[derive(Clone, Copy, Debug)]
+pub struct Slippage(f64);
+
+impl Slippage {
+    /// Creates a new slippage value. The value represents the slippage
+    /// tolerance as a factor (so 0.05 means that up to 5% slippage tolerace is
+    /// accepted, i.e. a swap can return 5% less than promised). Returns `None`
+    /// if an invalid value outside of the range [0.0, 1.0] is specified.
+    pub fn new(value: f64) -> Option<Self> {
+        if !(0.0..=1.0).contains(&value) {
+            return None;
+        }
+        Some(Self(value))
```

Comment:
No strong preference here, just wanted to spread awareness for `Option::then_some()`.
```suggestion
        (0.0..=1.0).contains(&value).then_some(Self(value))
```

Reply:
I always forget it exists. Awesome!